packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+        return [valueKey, unitKey];
+    }
+
+    public readonly cellTemplate: ViewTemplate<TableCellState<TableColumnNumberWithUnitCellData>> =
+        html<TableCellState<TableColumnNumberWithUnitCellData>>`
+            <nimble-text-field
+                readonly="true"
+                value=${x => formatData(x.data.value, x.data.units)}
+            >
+            </nimble-text-field>
+        `;
+
+    public validateCellData(cellData: TCellData): boolean {
+        return typeof cellData['value'] === 'number' && typeof cellData['units'] === 'string';
+    }
+}
+```
+
+Here we have a column visualized in different ways based on custom logic:
+
+```TS
+type TableColumnPositiveNegativeNumberCellData = NumberColumnData<'value'>;
+
+const isPositive = (value: number): bool => {
+    return value >= 0;
+}
+
+public TableColumnPositiveNegativeNumber extends FoundationElement implements ITableColumn<TableColumnPositiveNegativeNumberCellData> {
+    ...
+
+    public cellDataKeyNames = ['value'] as const;
+
+    @attr
+    public valueKey: string;
+
+    public getRowDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public readonly cellStyles: ElementStyles = 
+        css`
+            .good {
+                color: green;
+            }
+
+            .bad {
+                color: red;
+            }
+        `;
+
+    public readonly cellTemplate: ViewTemplate<TableCellState<TableColumnPositiveNegativeNumberCellData>> =
+        html<TableCellState<TableColumnPositiveNegativeNumberCellData>`
+            html`
+                <nimble-text-field
+                    class=${_ => isPositive(x.data.value) ? "good" : "bad"}
+                    readonly="true"
+                    value=${x => x.data.value}
+                    style="color: red;"
+                >
+                </nimble-text-field>
+            `}
+        `;
+
+    public validateCellData(cellData: TCellData): boolean {
+        return typeof cellData['value'] === 'number';
+    }
+}
+```
+
+Finally, here is a column that allows a user to register a callback for a click event on a button inside the cell template:
+
+```TS
+type TableColumnButtonCellData = StringColumnData<'id'>;
+
+const fireEvent = (buttonElement: HTMLElement, clickData: { data: string}): void => {
+    const event = new CustomEvent('button-click', { detail: data });
+    buttonElement.dispatchEvent(event);
+}
+
+public TableColumnButton extends FoundationElement implements ITableColumn<TableColumnButtonCellData> {
+    ...
+
+    public cellDataKeyNames = ['id'] as const;
+
+    @attr
+    public idKey: string;
+
+    public getRowDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public callback: (id: string) => void;
+
+    public readonly cellTemplate: ViewTemplate<TableCellState<TableColumnButtonCellData>> =
+        html<TableCellState<TableColumnButtonCellData>>`
+            <nimble-button @click="${(x, c) => fireEvent(c.event.currentTarget, {data: x.data.id})}>
```

Comment:
nit, your example is 100% accurate. We can simplify it a tad by knowing that this specific element (all fast-based elements) expose a [$emit](https://www.fast.design/docs/api/fast-element.fastelement._emit/#fastelementemit-method) method on the base class that does the same thing as the fireEvent function.

Makes the example simpler, but is the exact same concept.

```suggestion
            <nimble-button @click="${(x, c) => c.event.currentTarget.$emit('button-click'), {data: x.data.id})}">
```